### ISSUE TITLE
[selectors-4] additional resource state pseudo-classes for img / picture elements

### ISSUE THREAD

**Issue by voxpelli** [w3c/csswg-drafts#7467]
The [Resource State Pseudo-classes](https://drafts.csswg.org/selectors/#resource-pseudos) section mentions (emphasis mine):

> The pseudo-classes in this section apply to elements that represent loaded resources, particularly **images**/videos

Yet none of the currently defined pseudo classes seems to actually apply to any images.

[Following up](https://twitter.com/voxpelli/status/1545047038466830342) on [a recent popular tweet](https://twitter.com/jh3yy/status/1543700258018000896) that used `onerror` on `img`-tags to style failed images, I would like to mimic https://github.com/w3c/csswg-drafts/issues/3821 and resurrect a few of the pseudo-classes suggested in https://github.com/w3c/csswg-drafts/issues/3134

Primarily I would like to see:

* `:failed` – a replacement for the use of `onerror` to style images that fail to load
* `:loading` – a way to style images that are in the process of loading (would include both downloading and decoding I think)

Could also be useful to support the existing:

* `:stalled`– as images on bad internet connections can very much stall to load as well, especially a progressive JPEG which could me stuck in a low quality state.

**Comment by fantasai** [comment]
Makes sense to me. I wonder if :buffering and :loading should be interchangeable?

Also :failed should probably be :loading-failed, since lots of things can fail in a lot of different ways. :)

Agenda+ to ask the CSSWG what it wants to do here.

**Comment by 06b** [comment]
If it is any help for whoever creates the proposal, Firefox has the `:-moz-broken` pseudo-class which could be used as a model for broken images. This allows for the use case of setting a fallback via `content: url()` only when the image fails (as it currently always replaces an element as noted in #2831).

```
img:-moz-broken {
    content: url("placeholder.jpg");
}
```

**Comment by Crissov** [comment]
I’d like to suggest that HTTP status codes, particularly those commonly encountered in response to GET requests, be associated with some of the pseudo-classes proposed herein and in #3134. They may also inform the selection of some other states that may have frontend presentation, e.g.:

* `:partial` — 206, 416
* `:complete` — 200
* `:broken` — 404, 410
* `:failed` – 401, 402, 403, 406, …; 500
* `:loading` – 30x, before receiving 200 or 40x
* `:stalled` – 503, 504
* `:cached` — 304
* `:redirected` — 300, 301, 302, 307, 308 encountered in the request process, i.e. not mutually exclusive with most of the other pseudoclasses

**Comment by Mouvedia** [comment]
> * `:loading` – 30x, before receiving 200 or 40x

Something that might be worth discussing is whether it should copy `:-moz-loading` matching mechanism or not.
i.e. match elements which are waiting to be loaded

> [Note that images that are in the process of loading are not matched by this pseudo-class.](https://developer.mozilla.org/en-US/docs/Web/CSS/:-moz-loading)

**Comment by Crissov** [comment]
I guess there is also value in something like `:delayed` or `:scheduled` for lazy-loading scenarios.

**Comment by freedimension** [comment]
Came here to propose a `:loaded` or `:complete` pseudo attribute for **iframes** specifically, but then found this issue. Would be really nice to expose such states for all kinds of post loaded contents.

**Comment by SebastianZ** [comment]
I think @Crissov's [suggested pseudo-classes](https://github.com/w3c/csswg-drafts/issues/7467#issuecomment-1412653946) are a valid proposal. And seeing no objections on the suggestion so far, let's bring this to the agenda again.

Sebastian

**Comment by ascorbic** [comment]
_Context: I maintain [Unpic](https://github.com/ascorbic/unpic-img/), and built the official responsive image support in Gatsby and Astro._

I found this after @una asked framework authors for UI needs and went to look if it had been proposed. This would be very useful for implementing image loading placeholders. I like @Crissov's proposal, but there are some states that would not map to HTTP response codes. Most importantly `:loading` should also match images where a response has not been received yet. I'd also bikeshed `:loaded` as being better than `:complete`. 

It would also be helpful to a pseudo-class to match lazy-loaded images that are not being loaded, because they are offscreen or whatever (`:pending` maybe? `:offscreen`?)

**Comment by Crissov** [comment]
For large and complex pictures or weak devices, _loaded_ and _complete_ may not be the same thing because processing and rendering takes some time, too, but the difference might be negligible for all CSS authoring use cases.

**Comment by noamr** [comment]
> For large and complex pictures or weak devices, _loaded_ and _complete_ may not be the same thing because processing and rendering takes some time, too, but the difference might be negligible for all CSS authoring use cases. 

It should by and large follow the image load event. Async deciding can have some nuance here but it should be kept as a nuance.

**Comment by ascorbic** [comment]
I've been thinking a bit more about @Crissov's proposal and on second thoughts it does seem that a lot of those are redundant and can be subsumed into `:broken`. My suggestion would be the following:

- `:pending` – lazy-loaded images that have not yet started to load because the browser does not yet estimate that they will be needed
- `:loading` – a request has been sent, and has not errored or loaded
- `:stalled` –  the image is loading but has failed to receive data for some amount of time, but not failed. `:loading` will also match whenever `:stalled` matches. This pseudo-class is already used for video
- `:broken` – the request for the image has failed, whether from a HTTP error or timeout, or the image can't be used for some reason such as corrupt data. This would probably be when the `error` event has fired.
- `:loaded` – the image has sucessfully loaded. This would be when the `load` event has fired.

I don't know if there should be an additional `:complete` pseudo-class that would match both `:broken` and `:loaded`.

**Comment by Crissov** [comment]
Just for the record, existing [`:buffering` and `:stalled`](https://drafts.csswg.org/selectors-4/#media-loading-state) both imply `:playing` since they’re tailored to embedded content media with a temporal dimension, i.e. `<video>` and `<audio>`.

I now think that my mapping between `:stalled` and, for instance, codes 503 and 504 does not match the current definition of that pseudo-class.

**Comment by Th3S4mur41** [comment]
I wrote about [a use case where I need to detect when an image has successfully loaded to hide a fallback avatar with initials](https://th3s4mur41.me/blog/css-complete-pseudo-class/).

While I explicitly referenced `:complete` (post was based on https://github.com/w3c/csswg-drafts/issues/3134), the same result could be achieved with `img:not(:failed):not(:loading)`, as previously suggested. 

That makes `:complete` potentially redundant if `:loading` and `:failed` are well-defined.

Thanks for pushing this forward. These selectors are incredibly useful for clean, declarative UI logic.

**Comment by henke37** [comment]
whatwg/html#4806 has some related history worth reading up on.

**Comment by ascorbic** [comment]
If this is accepted, I'd like to help with the draft if that's possible. From the guidelines I think I need to find an Editor to work with?

**Comment by noamr** [comment]
A few issues:

1. One thing about some of these pseudo-class proposal, similar to what we had with `::picker(select)`. If we add them to images now, we can't use those names for other things in the future, because e.g. stylesheets would have `:loading` scattered around which would now gain new meaning. This is not a deal breaker, but perhaps it should be `:loading-image` or some such?

2. Feels like "non-OK status" and "broken" are different. e.g. if we've received a default image for our 404 request it's different from a corrupt request or invalid URL.
 
3. Another issue relating to the above is that we can't tell what the HTTP status code is for cross-origin no-cors images. This means this selector cannot differentiate between "successful" and "default image with a 404 response".

**Comment by noamr** [comment]
For loaded/loading, based on comments from @emilio, perhaps the useful values would be:
One that means "presenting a fully loaded image" and one that means "the current presented image matches (or doesnt match) the src attribute"

:fully-loaded and :stale?

Those can be combined for "up to date and fully loaded" etc.

**Comment by annevk** [comment]
There should be a corresponding HTML issue/PR for these pseudo-classes. Is anyone taking care of that? (See also https://github.com/whatwg/html/issues/9346 for an existing issue on making sure all this state is consistent. I suspect there are more.)

**Comment by svgeesus** [comment]
Selectors 4 is being stabilized to move to Recommendation. Proposed new features are being re-tagged for Selectors-5.

**Comment by jakearchibald** [comment]
I think it'd be worth having a pseudo-class to represent "an initial pass of the image is displayable" for progressive formats.
